ksmt-core/src/main/kotlin/org/ksmt/expr/rewrite/simplify/KFpExprSimplifier.kt

```diff
@@ -0,0 +1,696 @@
+                // todo: evaluate
+                return@simplifyApp mkFpToFpExpr(expr.sort, rm, value)
+            }
+
+            mkFpToFpExpr(expr.sort, rm, value)
+        }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : KFpSort> transform(expr: KRealToFpExpr<T>): KExpr<T> =
+        simplifyApp(expr as KApp<T, KExpr<KSort>>) { (rmArg, valueArg) ->
+            val rm: KExpr<KFpRoundingModeSort> = rmArg.uncheckedCast()
+            val value: KExpr<KRealSort> = valueArg.uncheckedCast()
+
+            if (rm is KFpRoundingModeExpr && value is KRealNumExpr) {
+                // todo: evaluate
+                return@simplifyApp mkRealToFpExpr(expr.sort, rm, value)
+            }
+
+            mkRealToFpExpr(expr.sort, rm, value)
+        }
+
+    override fun <T : KFpSort> transform(expr: KFpToRealExpr<T>): KExpr<KRealSort> = simplifyApp(expr) { (arg) ->
+        if (arg is KFpValue<T>) {
+            if (!arg.isNan() && !arg.isInfinity()) {
+                val decimalValue = fpDecimalValue(arg)
+                if (decimalValue != null) {
+                    val decimalPower = decimalValue.scale()
+                    var numerator = decimalValue.unscaledValue()
+                    var denominator = BigInteger.ONE
+                    if (decimalPower >= 0) {
+                        numerator *= BigInteger.TEN.pow(decimalPower)
+                    } else {
+                        denominator = BigInteger.TEN.pow(decimalPower.absoluteValue)
+                    }
+                    return@simplifyApp mkRealNum(numerator.expr, denominator.expr)
+                }
+            }
+        }
+        mkFpToRealExpr(arg)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : KFpSort> transform(expr: KBvToFpExpr<T>): KExpr<T> =
+        simplifyApp(expr as KApp<T, KExpr<KSort>>) { (rmArg, bvValueArg) ->
+            val rm: KExpr<KFpRoundingModeSort> = rmArg.uncheckedCast()
+            val value: KExpr<KBvSort> = bvValueArg.uncheckedCast()
+
+            if (rm is KFpRoundingModeExpr && value is KBitVecValue<*>) {
+                // todo: evaluate
+                return@simplifyApp mkBvToFpExpr(expr.sort, rm, value, expr.signed)
+            }
+
+            mkBvToFpExpr(expr.sort, rm, value, expr.signed)
+        }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : KFpSort> transform(expr: KFpToBvExpr<T>): KExpr<KBvSort> =
+        simplifyApp(expr as KApp<KBvSort, KExpr<KSort>>) { (rmArg, valueArg) ->
+            val rm: KExpr<KFpRoundingModeSort> = rmArg.uncheckedCast()
+            val value: KExpr<T> = valueArg.uncheckedCast()
+
+            if (rm is KFpRoundingModeExpr && value is KFpValue<T>) {
+                if (!value.isNan() && !value.isInfinity()) {
+                    val decimalValue = fpDecimalValue(value)
+                    if (decimalValue != null) {
+                        val lowLimit = if (expr.isSigned) {
+                            -BigDecimal.valueOf(2).pow(expr.bvSize - 1)
```

Comment:
Maybe write clarifying parenthesis to avoid operation application priorities misunderstanding?